src/FFNetAgent.cpp

```diff
@@ -0,0 +1,365 @@
+                {GEOPM_DOMAIN_PACKAGE, "GEOPM_CPU_NN_PATH"},
+                {GEOPM_DOMAIN_GPU, "GEOPM_GPU_NN_PATH"}
+            };
+
+    const std::map<geopm_domain_e, std::string> FFNetAgent::M_FREQMAP_ENVNAME =
+            {
+                {GEOPM_DOMAIN_PACKAGE, "GEOPM_CPU_FMAP_PATH"},
+                {GEOPM_DOMAIN_GPU, "GEOPM_GPU_FMAP_PATH"}
+            };
+
+    const std::map<geopm_domain_e, std::string> FFNetAgent::M_MAX_FREQ_SIGNAL_NAME = 
+            {
+                {GEOPM_DOMAIN_PACKAGE, "CPU_FREQUENCY_MAX_AVAIL"},
+                {GEOPM_DOMAIN_GPU, "GPU_CORE_FREQUENCY_MAX_AVAIL"}
+            };
+
+    const std::map<geopm_domain_e, std::string> FFNetAgent::M_MIN_FREQ_SIGNAL_NAME = 
+            {
+                {GEOPM_DOMAIN_PACKAGE, "CPU_FREQUENCY_MIN_AVAIL"},
+                {GEOPM_DOMAIN_GPU, "GPU_CORE_FREQUENCY_MIN_AVAIL"}
+            };
+
+    const std::map<geopm_domain_e, std::string> FFNetAgent::M_MAX_FREQ_CONTROL_NAME = 
+            {
+                {GEOPM_DOMAIN_PACKAGE, "CPU_FREQUENCY_MAX_CONTROL"},
+                {GEOPM_DOMAIN_GPU, "GPU_CORE_FREQUENCY_MAX_CONTROL"}
+            };
+    const std::map<geopm_domain_e, std::string> FFNetAgent::M_MIN_FREQ_CONTROL_NAME = 
+            {
+                {GEOPM_DOMAIN_PACKAGE, "CPU_FREQUENCY_MIN_CONTROL"},
+                {GEOPM_DOMAIN_GPU, "GPU_CORE_FREQUENCY_MIN_CONTROL"}
+            };
+
+    const std::map<geopm_domain_e, std::string> FFNetAgent::M_TRACE_SUFFIX = 
+            {
+                {GEOPM_DOMAIN_PACKAGE, "_cpu_"},
+                {GEOPM_DOMAIN_GPU, "_gpu_"}
+            };
+
+
+    FFNetAgent::FFNetAgent()
+        : FFNetAgent(platform_io(), platform_topo(), {}, {})
+    {
+
+    }
+
+    FFNetAgent::FFNetAgent(
+            PlatformIO &plat_io,
+            const PlatformTopo &topo,
+            const std::map<std::pair<geopm_domain_e, int>, std::shared_ptr<DomainNetMap> >
+                &net_map,
+            const std::map<geopm_domain_e, std::shared_ptr<RegionHintRecommender> >
+                &freq_recommender
+            )
+        : m_platform_io(plat_io)
+          , m_last_wait{{0, 0}}
+          , m_do_write_batch(false)
+          , m_perf_energy_bias(0)
+    {
+        geopm_time(&m_last_wait);
+        init_domain_indices(topo);
+
+        if (freq_recommender.empty()) {
+            for (geopm_domain_e domain_type : m_domain_types) {
+                m_freq_recommender[domain_type] = RegionHintRecommender::make_shared(
+                        get_env_value(M_FREQMAP_ENVNAME.at(domain_type)),
+                        m_platform_io.read_signal(
+                            M_MIN_FREQ_SIGNAL_NAME.at(domain_type),
+                            GEOPM_DOMAIN_BOARD,
+                            0),
+                        m_platform_io.read_signal(
+                            M_MAX_FREQ_SIGNAL_NAME.at(domain_type),
+                            GEOPM_DOMAIN_BOARD,
+                            0)
+                        );
```

Comment:
This block and several below have a different indentation style that what we normally use.  Please fix.  I can help on this.